File: 1255. Maximum Score Words Formed by Letters.py
class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        count = [0] * 26
        for letter in letters:
            count[ord(letter)-ord('a')] += 1
        return self.find(words, count, score, 0)
    def find(self, words, count, score, index):
        if index >= len(words): return 0
        answer = self.find(words, count, score, index+1)
        if self.can_make(words[index], count):
            for letter in words[index]:
                count[ord(letter)-ord('a')] -= 1
            word_score = self.find_score(words[index], score)
            answer = max(answer, word_score + self.find(words, count, score, index+1))
            for letter in words[index]:
                count[ord(letter)-ord('a')] += 1
        return answer
    def can_make(self, word, count):
        word_count = [0] * 26
        for letter in word:
            word_count[ord(letter)-ord('a')] += 1
        for i in range(26):
            if count[i] < word_count[i]: return False
        return True
    def find_score(self, word, score):
        answer = 0
        for letter in word:
            answer += score[ord(letter)-ord('a')]
        return answer

File: 1255. Maximum Score Words Formed by Letters.go
func maxScoreWords(words []string, letters []byte, score []int) int {
	count := make([]int, 26, 26)
	for _, letter := range letters {
		count[letter-'a'] += 1
	}
	return find(words, count, score, 0)
}

func find(words []string, count, score []int, index int) int {
	if index >= len(words) { return 0 }
	answer := find(words, count, score, index+1)
	if canMake(words[index], count) {
		makeWord(words[index], count)
		wordScore := findScore(words[index], score)
		answer = max(answer, wordScore + find(words, count, score, index+1))
		unmakeWord(words[index], count)
	}
	return answer
}

func findScore(word string, score []int) int {
	answer := 0
	for _, letter := range word {
		answer += score[letter-'a']
	}
	return answer
}

func canMake(word string, count []int) bool {
	wordCount := make([]int, 26, 26)
	for _, letter := range word {
		wordCount[letter-'a'] += 1
	}
	for i := range 26 {
		if wordCount[i] > count[i] { return false }
	}
	return true
}

func makeWord(word string, count []int) {
	for _, letter := range word {
		count[letter - 'a'] -= 1
	}
}

func unmakeWord(word string, count []int) {
	for _, letter := range word {
		count[letter - 'a'] += 1
	}
}